lib/python/qmk/keymap.py

```diff
@@ -98,3 +101,57 @@ def write(keyboard, keymap, layout, layers):
         keymap_fd.write(keymap_c)
 
     return keymap_file
+
+def find_keymaps(base_path, revision = "", community = False):
+    """ Find the available keymaps for a keyboard and revision pair.
+
+    Args:
+        base_path: The base path of the keyboard.
+
+        revision: The keyboard's revision.
+
+        community: Set to True for the layouts under layouts/community.
+
+    Returns:
+        a list with whe keymaps's names
+    """
+    path_wildcard = os.path.join(base_path, "**", "keymap.c")
+    if community:
+        path_regex = re.compile(r"^" + re.escape(base_path) + "(\S+)" + os.path.sep + "keymap\.c")
+    else:
+        path_regex = re.compile(r"^" + re.escape(base_path) + "(?:" + re.escape(revision) + os.path.sep + ")?keymaps" + os.path.sep + "(\S+)" + os.path.sep + "keymap\.c")
+    names = [path_regex.sub(lambda name: name.group(1), path) for path in glob.iglob(path_wildcard, recursive = True) if path_regex.search(path)]
+    return names
+
+def list_keymaps(keyboard_name):
+    """ List the available keymaps for a keyboard.
+
+    Args:
+        keyboard_name: the keyboards full name with vendor and revision if necessary, example: clueboard/66/rev3
+
+    Returns:
+        a list with the names of the available keymaps
+    """
+    if os.path.sep in keyboard_name:
+        keyboard, revision = os.path.split(os.path.normpath(keyboard_name))
+    else:
+        keyboard = keyboard_name
+        revision = ""
+
+    # parse all the rules.mk files for the keyboard
+    rules_mk = qmk.makefile.get_rules_mk(keyboard, revision)
+    names = list()
+
+    if rules_mk:
+        # get the keymaps from the keyboard's directory
+        kb_base_path = os.path.join(os.getcwd(), "keyboards", keyboard) + os.path.sep
+        names = find_keymaps(kb_base_path, revision)
+
+        # if community layouts are supported, get them
+        if "LAYOUTS" in rules_mk:
+            for layout in rules_mk["LAYOUTS"]["value"].split():
+                cl_base_path = os.path.join(os.getcwd(), "layouts", "community", layout) + os.path.sep
+                names = names + find_keymaps(cl_base_path, revision, community = True)
```

Comment:
I'd like to see this more closely match what happens in Make. To do that here's what we need to do:

1. Gather all keymaps from supported community layouts
2. Move up the directory tree until you encounter `qmk_firmware/keyboards`
3. Move back down the directory tree checking for the existence of a `keymaps` directory.
4. Whenever `keymaps` is encountered add these keymaps to the list that will be returned. If you are tracking the keymap's directory overwrite any already existing keymaps (last match wins).

The resulting list from those heuristics should be the full list of keymaps for a given keyboard.

As a bonus I think this will remove the need for the regex above.

Reply:
I'm doing basically that, just starting with the `keymaps` directories and after the community layouts. In this context it shouldn't matter, as the unique'd end result should be the same. 
But thanks for reminding me to unique the keymaps, I totally forgot that.

I will look into how to optimize this process.